core/imds/src/main/java/software/amazon/awssdk/imds/MetadataResponse.java

```diff
@@ -0,0 +1,187 @@
+
+/**
+ * The class is used for Response Handling and Parsing the metadata fetched by the get call in the {@link Ec2Metadata} interface.
+ * Metadata is stored in the instance variable <b>body</b>. The class provides convenience methods to the users to parse the
+ * metadata as a String, List and to parse the metadata in the JsonResponse according to the key.
+ */
+@SdkPublicApi
+public class MetadataResponse {
+
+    private static final Logger log = LoggerFactory.getLogger(MetadataResponse.class);
+
+    private static final JsonNodeParser JSON_NODE_PARSER = JsonNode.parser();
+    
+    private final String body;
+
+    public MetadataResponse(String body) {
+        this.body = body;
+    }
+
+    /**
+     * Returns the Metadata Response body as a String. This method can be used for parsing the retrieved
+     * singular metadata from IMDS.
+     *
+     * @return String Representation of the Metadata Response Body.
+     *
+     * <p>
+     * Example:
+     * <pre>
+     * {@code
+     *
+     * Ec2Metadata ec2Metadata = Ec2Metadata.builder().build();
+     * MetadataResponse metadataResponse = client.get("/latest/meta-data/ami-id");
+     * String response = metadataResponse.asString();
+     *  }
+     *  </pre>
+     */
+    public String asString() {
+        return body;
+
+    }
+
+    /**
+     * Parses the response String into a list of Strings split by delimiter ("\n"). This method can be used for parsing the
+     * list-type metadata from IMDS.
+     *
+     * @return List Representation of the Metadata Response Body.
+     *
+     * <p>
+     * Example:
+     * <pre>
+     * {@code
+     *
+     * Ec2Metadata ec2Metadata = Ec2Metadata.builder().build();
+     * MetadataResponse metadataResponse = client.get("/latest/meta-data/ancestor-ami-ids");
+     * List<String>response = metadataResponse.asList();
+     * }
+     * </pre>
+     */
+    public List<String> asList() {
+
+        if (null != body && body.contains("\n"))  {
+
+            return Arrays.asList(body.split("\n"));
+        }
+
+        return Collections.emptyList();
+    }
+
+    /**
+     * Parses the response String to get the String Array Values of the key in jsonResponse. This method can be used for
+     * parsing the metadata in a String Json Format.
+     *
+     * @param key The key / field to retrieve from jsonResponse.
+     * @return String Array of the values of the key in the jsonResponse.
+     * Returns Null in case the key is null or key doesn't exist in the json.
+     *
+     * <p>
+     * Example:
+     * <pre>
+     * {@code
+     *
+     * Ec2Metadata ec2Metadata = Ec2Metadata.builder().build();
+     * MetadataResponse metadataResponse = client.get("/latest/dynamic/instance-identity/document");
+     * String[] devpayProductCodes = metadataResponse.getStringArrayValuesFromJson("devpayProductCodes");
+     * }
+     * </pre>
+     */
+
+    public String[] getStringArrayValuesFromJson(String key) {
```

Comment:
Instead of having these json-like methods, maybe we should just have one to parse the body into a `software.amazon.awssdk.core.document.Document`? That would require making some currently-internal stuff protected (parsing JSON into a document), but it would be more flexible.

E.g.
```java
Ec2MetadataClient ec2Metadata = Ec2MetadataClient.create();
MetadataResponse metadataResponse = client.get("/latest/dynamic/instance-identity/document");
Document instanceIdentity = metadataResponse.asDocument();
```

Reply:
Great insight. Will implement it.